File: phpServer/tilemill.php
<?php

/**
 * Main bootstrap function. Inits configuration and routes requests to
 * respective request handlers.
 */
function main($path, $args, $method = 'get') {
  $config = array('files' => dirname(__FILE__));
  if (file_exists('tilemill.cfg')) {
    $config = array_merge($config, parse_ini_file('tilemill.cfg'));
  }

  switch ($path) {
    case '':
      $handler = new InfoHandler($config, $path, $args);
      return $handler->{$method}();
    case 'list':
      $handler = new ListHandler($config, $path, $args);
      return $handler->{$method}();
    case 'add':
      $handler = new AddHandler($config, $path, $args);
      return $handler->{$method}();
    case 'file':
      $handler = new FileHandler($config, $path, $args);
      return $handler->{$method}();
    default:
      $handler = new NotFoundHandler($config, $path, $args);
      return $handler->{$method}();
  }
}

/**
 * Base request handler class.
 */
class RequestHandler {
  var $config;
  var $path;
  var $args;

  function __construct($config, $path, $args) {
    $this->config = $config;
    $this->path = $path;
    $this->args = $args;
  }

  function get() {
    header('HTTP/1.1 200 OK');
    header('Connection: close');
    return;
  }

  function post() {
    header('HTTP/1.1 200 OK');
    header('Connection: close');
    return;
  }

  function json($data, $force_json = TRUE) {
    if (isset($this->args['jsoncallback'])) {
      header('Content-Type: text/javascript');
      $json = json_encode($data);
      return "{$this->args['jsoncallback']}({$json})";
    }
    elseif ($force_json) {
      return json_encode($data);
    }
    return $data;
  }

  function safePath($path) {
    return strpos($path, '..') === FALSE && !preg_match('/[^\w+\/._-]/', $path);
  }
}

/**
 * Info handler class.
 * @TODO: determine what other information we need to provide.
 */
Class InfoHandler extends RequestHandler {
  function get() {
    parent::get();
    return $this->json(array(
      'api' => 'basic',
      'version' => 1.0,
    ));
  }
}

/**
 * 404 handler class.
 */
class NotFoundHandler extends RequestHandler {
  function get() {
    header('HTTP/1.1 404 Not Found');
    header('Connection: close');
    return;
  }
  function post() {
    header('HTTP/1.1 404 Not Found');
    header('Connection: close');
    return;
  }
}

/**
 * List handler class.
 */
class ListHandler extends RequestHandler {
  function get() {
    parent::get();
    if (isset($this->args['filename'])) {
      $path = "{$this->config['files']}/{$this->args['filename']}";
      if ($this->safePath($path)) {
        if (file_exists($path) && is_dir($path) && $handle = opendir($path)) {
          $directories = array();
          while (FALSE !== ($file = readdir($handle))) {
            $basename = basename($file);
            if (file_exists("{$path}/{$file}/{$basename}.mml") && is_file("{$path}/{$file}/{$basename}.mml")) {
              $directories[] = $basename;
            }
          }
          return $this->json(array('status' => TRUE, 'data' => $directories));
        }
        return $this->json(array('status' => FALSE, 'data' => 'The file could not be found'));
      }
    }
    return $this->json(array('status' => FALSE, 'data' => 'Invalid filename'));
  }
}

/**
 * Add handler class.
 */
class AddHandler extends RequestHandler {
  function post() {
    parent::post();
    if (isset($this->args['filename'])) {
      $path = "{$this->config['files']}/{$this->args['filename']}";
      if ($this->safePath($path)) {
        if (!file_exists($path)) {
          mkdir($path);
          return $this->json(array('status' => TRUE), TRUE);
        }
        return $this->json(array('status' => FALSE, 'data' => "The directory {$path} already exists"));
      }
    }
    return $this->json(array('status' => FALSE, 'data' => 'Invalid filename'));
  }
}

/**
 * File handler class.
 */
class FileHandler extends RequestHandler {
  function get() {
    parent::get();
    if (isset($this->args['filename'])) {
      $path = "{$this->config['files']}/{$this->args['filename']}";
      if ($this->safePath($path)) {
        if (file_exists($path) && is_file($path)) {
          $data = file_get_contents($path);
          return $this->json($data, FALSE);
        }
        return $this->json(array('status' => FALSE, 'data' => 'The file could not be found'));
      }
    }
    return $this->json(array('status' => FALSE, 'data' => 'Invalid filename'));
  }
  function post() {
    parent::post();
    if (isset($this->args['filename'])) {
      $path = "{$this->config['files']}/{$this->args['filename']}";
      if ($this->safePath($path)) {
        // Create paths in between.
        if (!file_exists(dirname($path))) {
          mkdir(dirname($path), 0777, TRUE);
        }
        // Write file.
        if (file_exists(dirname($path)) && is_dir(dirname($path))) {
          file_put_contents($path, $this->args['data']);
          return $this->json(array('status' => TRUE));
        }
        return $this->json(array('status' => FALSE, 'data' => "Could not write file"));
      }
    }
    return $this->json(array('status' => FALSE, 'data' => 'Invalid filename'));
  }
}

$method = strtolower($_SERVER['REQUEST_METHOD']);
$args = $method === 'post' ? $_POST : $_GET;
print main($_GET['q'], array_diff_key($args, array('q' => NULL)), $method);
exit;

File: pythonServer/tilemill.py
#!/usr/bin/env python

import os.path
import unicodedata
import tornado.httpserver
import tornado.ioloop
import tornado.options
import tornado.web
import tornado.escape
import tornado.template
import re
import shutil

from urlparse import urlparse
from tornado.options import define, options

define("port",
        default=8889,
        help="run on the given port",
        type=int)
define("files",
        default=os.path.dirname(__file__),
        help="files directory",
        type=str)
define("config",
        default=os.path.join(os.path.dirname(__file__), "tilemill.cfg"),
        help="path to configuration file",
        type=str)

class TileMill(tornado.web.RequestHandler):
    def json(self, json, force_json = False):
        """ serve a page with an optional jsonp callback """
        if self.get_argument('jsoncallback', None):
            json = tornado.escape.json_encode(json)
            json = "%s(%s)" % (self.get_argument('jsoncallback', None), json)
            self.set_header('Content-Type', 'text/javascript')
        elif force_json:
            json = tornado.escape.json_encode(json)
        self.write(json)
    def safePath(self, path):
        return True or path.find('..') == -1 and not re.search('[^\w.-_\/]', path)

class InfoHandler(TileMill):
    def get(self):
        self.json({ 'api': 'basic', 'version': 1.0 }, True)

class ListHandler(TileMill):
    def get(self):
        path = os.path.join(options.files, self.get_argument('filename'))
        if (self.safePath(path) and os.path.isdir(path)):
            directories = []
            for root, dirs, files in os.walk(path):
                basename = os.path.basename(root)
                if os.path.isfile(os.path.join(root, basename + '.mml')):
                    directories.append(basename)
            self.json({
                'status': True,
                'data': directories
                }, True)
        elif (self.safePath(path)):
            self.json({
                'status': False,
                'data': 'The file could not be found'
                }, True)
        else:
            self.json({
                'status': False,
                'data': 'Invalid filename'
                }, True)

class MtimeHandler(TileMill):
    def get(self):
        path = os.path.join(options.files, self.get_argument('filename'))
        if (self.safePath(path) and os.path.isfile(path)):
            mtime = os.path.getmtime(path)
            self.json({
                'status': True,
                'mtime': mtime,
                'filename': self.get_argument('filename')
                }, False)
        elif (self.safePath(path)):
            self.json({
                'status': False,
                'data': 'The file could not be found'
                }, True)
        else:
            self.json({
                'status': False,
                'data': 'Invalid filename'
                }, True)

class FileHandler(TileMill):
    def get(self):
        path = os.path.join(options.files, self.get_argument('filename'))
        if (self.safePath(path) and os.path.isfile(path)):
            buffer = open(path)
            data = buffer.read()
            buffer.close()
            self.json(data, False)
        elif (self.safePath(path)):
            self.json({
                'status': False,
                'data': 'The file could not be found'
                }, True)
        else:
            self.json({
                'status': False,
                'data': 'Invalid filename'
                }, True)

    def post(self):
        path = os.path.join(options.files, self.get_argument('filename'))
        if (self.safePath(path)):
            method = self.get_argument('method', 'put')
            data = self.get_argument('data', '')
            if method == 'delete':
                self.rm(path);
                self.json({
                    'status': True
                    }, True)
            else:
                if (not os.path.isdir(os.path.dirname(path))):
                    os.makedirs(os.path.dirname(path))
                if (os.path.isdir(os.path.dirname(path))):
                    buffer = open(path, 'w')
                    buffer.writelines(data)
                    buffer.close()
                    self.json({
                        'status': True
                        }, True)
                else:
                    self.json({
                        'status': False,
                        'data': 'Could not write file'
                        }, True)
        elif (self.safePath(path)):
            self.json({
                'status': False,
                'data': 'Could not write file'
                }, True)
        else:
            self.json({
                'status': False,
                'data': 'Invalid filename'
                }, True)

    def rm(self, path):
        for root, dirs, files in os.walk(path, topdown=False):
            for name in files:
                os.remove(os.path.join(root, name))
            for name in dirs:
                os.rmdir(os.path.join(root, name))

class Application(tornado.web.Application):
    def __init__(self):
        handlers = [
            (r"/", InfoHandler),
            (r"/list", ListHandler),
            (r"/file", FileHandler),
            (r"/mtime", MtimeHandler),
        ]
        settings = dict(
            template_path=os.path.join(os.path.dirname(__file__), "templates"),
            static_path=os.path.join(os.path.dirname(__file__), "static"),
        )
        tornado.web.Application.__init__(self, handlers, **settings)

def main():
    tornado.options.parse_config_file(options.config)
    tornado.options.parse_command_line()

    http_server = tornado.httpserver.HTTPServer(Application())
    http_server.listen(options.port)
    tornado.ioloop.IOLoop.instance().start()

if __name__ == "__main__":
    main()
